enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,237 @@
+- As a cluster admin, if a CSV is stuck in the FAILED state and I have enabled "fail forward" upgrades, I want to OLM allow the operator to upgrade to the next version.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Opting into Fail Forward Upgrades
+
+Since resolution is namespaced, the toggle for allowing "fail forward" upgrades should be namespaced as well to avoid having to add the extra burden on the resolver surrounding understanding which dependent operators can or cannot use the forced upgrade toggle.
+
+With this in mind, this enhancement proposes adding the `upgradeStrategy` field to an already watched namespaced resource, the OperatorGroup.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
+metadata:
+  name: foo
+  namespace: bar
+spec:
+  upgradeStrategy:
+     # Possible values include "Default" or "UnsafeFailForward".
+    name: UnsafeFailForward
+```
+
+With the the upgradeStrategy type is set to `UnsafeFailForward`, OLM will allow operators to "fail forward" in adherence with the principles discussed below. If the upgradeStrategy is unset or set to `Default` OLM will exhibit existing behavior. It is critical that the documentation surrounding `UnsafeFailForward` makes it clear to users that critical versions of operators may not be installed which can lead to unknown behavior and potentially data loss.
+
+#### Supporting "Fail Forward" Upgrades
+
+Before diving into the specific behavior OLM will need to support when "failing forward" from failed CSVs or InstallPlans, it is important to understand how OLM and the Resolver calculate what needs to be installed. When determining what to install, OLM provides the resolver with the set of CSVs in a namespace. The resolver treats these CSVs differently depending on whether or not they are claimed by a Subscription, where the Subscription lists the CSV's name in its `.status.currentCSV` field.
+- If a CSV is claimed by a Subscription, the resolver will allow it to be upgraded by a bundle that replaces it.
+- If a CSV is not claimed by a Subscription, it is assumed that the user installed the CSV themselves, that OLM should not upgrade the CSV, and that it must appear in the next set of installables.
+
+OLM's Resolver is deterministic, meaning that the set of installables will not change for a given set of arguments if no new upgrades have been declared in the existing CatalogSources. When an upgrade fails due to a CSV entering the FAILED phase, the CSV being replaced still exists and is not referenced by a subscription. Today, OLM would send both `Operator v1` and `Operator v2` to the resolver which will normally be unsatisfiable because `Operator v1` is marked as required and `Operator v2` cannot be upgraded further since it provides the same APIs as `Operator v1`. In order to support "failing forward", OLM will need to omit CSVs in the REPLACING phase from the set of arguments sent to the resolver when "fail forward" upgrades are enabled.
```

Comment:
non-blocking suggestion: [a mermaid graph](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/) here may be useful in describing the exact phenomenon here.

Reply:
This is a good idea, but I'm a bit strapped for time right now and may add it at a future point in time.